src/ir_Kelon.cpp

```diff
@@ -0,0 +1,500 @@
+              repeat, 50);
+}
+
+#endif  // SEND_KELON
+
+#if DECODE_KELON
+/// Decode the supplied Kelon message.
+/// Status: STABLE / Working.
+/// @param[in,out] results Ptr to the data to decode & where to store the result
+/// @param[in] offset The starting index to use when attempting to decode the
+///   raw data. Typically/Defaults to kStartOffset.
+/// @param[in] nbits The number of data bits to expect.
+/// @param[in] strict Flag indicating if we should perform strict matching.
+/// @return True if it can decode it, false if it can't.
+
+bool IRrecv::decodeKelon(decode_results *results, uint16_t offset,
+                         const uint16_t nbits, const bool strict) {
+  if (strict && nbits != kKelonBits) {
+    return false;
+  }
+  if (!matchGeneric(results->rawbuf + offset, results->state,
+                    results->rawlen - offset, nbits,
+                    kKelonHdrMark, kKelonHdrSpace,
+                    kKelonBitMark, kKelonOneSpace,
+                    kKelonBitMark, kKelonZeroSpace,
+                    kKelonBitMark, 0, false,
+                    _tolerance, 0, false)) {
+    return false;
+  }
+
+  results->decode_type = decode_type_t::KELON;
+  results->bits = nbits;
+  return true;
+}
+
+#endif  // DECODE_KELON
+
+/// Class constructor
+/// @param[in] pin GPIO to be used when sending.
+/// @param[in] inverted Is the output signal to be inverted?
+/// @param[in] use_modulation Is frequency modulation to be used?
+IRKelonAc::IRKelonAc(const uint16_t pin, const bool inverted,
+                     const bool use_modulation)
+    : _irsend{pin, inverted, use_modulation}, _{} { stateReset(); }
+
+/// Reset the internals of the object to a known good state.
+void IRKelonAc::stateReset() {
+  _.raw = 0L;
+  _.preamble[0] = 0b10000011;
+  _.preamble[1] = 0b00000110;
+}
+
+#if SEND_KELON
+
+/// Send the current internal state as an IR message.
+/// @param[in] repeat Nr. of times the message will be repeated.
+void IRKelonAc::send(const uint16_t repeat) {
+  _irsend.sendKelon(getRaw(), kKelonBits, repeat);
+
+  // Reset toggle flags
+  _.PowerToggle = false;
+  _.SwingVToggle = false;
+
+  // Remove the timer time setting
+  _.TimerHours = 0;
+  _.TimerHalfHour = 0;
+}
+
+/// Ensures the AC is on or off by exploiting the fact that setting
+/// it to "smart" will always turn it on if it's off.
+/// This method will send 2 commands to the AC to do the trick
+/// @param[in] on Whether to ensure the AC is on or off
+void IRKelonAc::ensurePower(bool on) {
+  // Try to avoid turning on the compressor for this operation.
+  // "Dry grade", when in "smart" mode, acts as a temperature offset that
+  // the user can configure if they feel too cold or too hot. By setting it
+  // to +2 we're setting the temperature to ~28°C, which will effectively
+  // set the AC to fan mode.
+  int8_t previousDry = getDryGrade();
+  setDryGrade(2);
+  setMode(kKelonModeSmart);
+  send();
+
+  _irsend.space(kKelonGap);
```

Comment:
Curious. Why is this needed? `send()` should be ending a message with `kKelonGap` already if I'm reading it correctly. If it requires a longer gap (i.e. 2 x `kDefaultMessageGap`) then you should probably make `kKelonGap` longer.

Reply:
I must have misread `sendGeneric`.